nav2_map_server/src/map_saver.cpp

```diff
@@ -1,250 +1,198 @@
 
-    std::string mode_str = declare_parameter("map_mode", "trinary");
-    try {
-      map_mode = map_mode_from_string(mode_str);
-    } catch (std::invalid_argument &) {
-      map_mode = MapMode::Trinary;
-      RCLCPP_WARN(
-        get_logger(), "Map mode parameter not recognized: '%s', using default value (trinary)",
-        mode_str.c_str());
-    }
+  save_map_timeout_ = std::make_shared<rclcpp::Duration>(
+    std::chrono::milliseconds(declare_parameter("save_map_timeout", 2000)));
 
-    image_format = declare_parameter("image_format", map_mode == MapMode::Scale ? "png" : "pgm");
-    std::transform(
-      image_format.begin(), image_format.end(), image_format.begin(),
-      [](unsigned char c) {return std::tolower(c);});
-    const std::vector<std::string> BLESSED_FORMATS{"bmp", "pgm", "png"};
-    if (
-      std::find(BLESSED_FORMATS.begin(), BLESSED_FORMATS.end(), image_format) ==
-      BLESSED_FORMATS.end())
-    {
-      std::stringstream ss;
-      bool first = true;
-      for (auto & format_name : BLESSED_FORMATS) {
-        if (!first) {
-          ss << ", ";
-        }
-        ss << "'" << format_name << "'";
-        first = false;
-      }
-      RCLCPP_WARN(
-        get_logger(), "Requested image format '%s' is not one of the recommended formats: %s",
-        image_format.c_str(), ss.str().c_str());
-    }
-    const std::string FALLBACK_FORMAT = "png";
-
-    try {
-      Magick::CoderInfo info(image_format);
-      if (!info.isWritable()) {
-        RCLCPP_WARN(
-          get_logger(), "Format '%s' is not writable. Using '%s' instead",
-          image_format.c_str(), FALLBACK_FORMAT.c_str());
-        image_format = FALLBACK_FORMAT;
-      }
-    } catch (Magick::ErrorOption & e) {
-      RCLCPP_WARN(
-        get_logger(), "Format '%s' is not usable. Using '%s' instead:\n%s",
-        image_format.c_str(), FALLBACK_FORMAT.c_str(), e.what());
-      image_format = FALLBACK_FORMAT;
-    }
-    if (
-      map_mode == MapMode::Scale &&
-      (image_format == "pgm" || image_format == "jpg" || image_format == "jpeg"))
-    {
-      RCLCPP_WARN(
-        get_logger(),
-        "Map mode 'scale' requires transparency, but format '%s' does not support it. Consider "
-        "switching image format to 'png'.",
-        image_format.c_str());
-    }
+  free_thresh_default_ = declare_parameter("free_thresh_default", 0.25),
+  occupied_thresh_default_ = declare_parameter("occupied_thresh_default", 0.65);
+}
 
-    RCLCPP_INFO(get_logger(), "Waiting for the map");
-    map_sub_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
-      "map", rclcpp::SystemDefaultsQoS(),
-      std::bind(&MapSaver::mapCallback, this, std::placeholders::_1));
-  }
+MapSaver::~MapSaver()
+{
+  RCLCPP_INFO(get_logger(), "Destroying");
 }
 
-void MapSaver::try_write_map_to_file(const nav_msgs::msg::OccupancyGrid & map)
+nav2_util::CallbackReturn
+MapSaver::on_configure(const rclcpp_lifecycle::State & /*state*/)
```

Comment:
Its not ideal that the CLI doesn't call on the lifecycle transitions (which I suppose it doesn't need to since it doesn't need a service) so a lot of the code is server specific, but I think we should go with it. 

No action item.

Reply:
Ok. Just some thoughts. Maybe it is not a straightforward solution to use LifecycleNode node both ways: as usual lifecycle node with manager changing its states (as `map_saver_server` does) and also as a just usual class utilizing some functionality from LifecycleNode (as `map_saver_cli` uses). We could make `map_saver_cli` to be started with LifecycleNode manager with changing its state to `activated->deactivated` after job is done. But this will only complicate the picture and will not bring anything new into CLI application. It is also seems to be not good idea to separate `MapSaver` into two CLI-dependent and server-dependent classes since there will be some duplicating code (like `saveMapTopicToFile()` method).

So I wonder what else can we come up with for future changes (after this PR) to make this to to use everything for its intended purpose without complication of the architecture?

Reply:
I think its fine as is. I don’t think the other alternatives are somehow clearer, more superior, or worth the effort.